src/functional/swing_slide_filter.zig

```diff
@@ -0,0 +1,552 @@
+// Copyright 2024 TerseTS Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implementation of Swing and Slide Filter algorithms from the paper:
+//! Hazem Elmeleegy, Ahmed K. Elmagarmid, Emmanuel Cecchet, Walid G. Aref, and Willy Zwaenepoel.
+//! Online piece-wise linear approximation of numerical streams with precision guarantees.
+//! Proc. VLDB Endow. 2, 1, 2009.
+//! https://doi.org/10.14778/1687627.1687645.
+
+const std = @import("std");
+const ts = @import("../tersets.zig");
+
+const mem = std.mem;
+const testing = std.testing;
+
+const ArrayList = std.ArrayList;
+
+const Segment = struct {
+    start_time: usize,
+    start_value: f64,
+    end_time: usize,
+    end_value: f64,
+};
+
+/// Line structure uses f80 to avoid overflows with f64
+const Line = struct {
+    slope: f80,
+    intercept: f80,
+};
+
+pub fn isWithinErrorBound(
+    uncompressed_values: []f64,
+    decompressed_values: []f64,
+    error_bound: f32,
+) bool {
+    for (decompressed_values, 0..) |item, i| {
+        if (@abs(uncompressed_values[i] - item) > error_bound + 1e-7) return false;
```

Comment:
This is now removed. A solution to avoid exceeding the error bound due to large floating points manipulation is to decrease the error bound before finding the approximation. However, the question raises, by how much should we reduce the error bound.